### DISCUSSION TITLE
The result of json is_number_integer() function is wrong when read a json file

### DISCUSSION THREAD

**Discussion by Ymxia1993** [nlohmann/json#2654] [Q&A]
code:
```C++
#include <fstream>
#include <iostream>
#include "nlohmann/json.hpp"
using nlohmann::json;
int main()
{
  std::ifstream file("Test.json");
  json j = json::parse(file);
  std::cout<<j["a"].is_number_integer()<<std::endl;
}
```
 json file:
`{
    "a":1
}`
when use icc compiler(**icc (ICC) 19.1.0.139 20191004**) compile code,  the result of json is_number_integer() function  is 0, its obviously a wrong result.
However, when I use gcc compiler the result is right.

Platform :Linux version 3.10.0-1062.1.2.el7.x86_64 (mockbuild@kbuilder.bsys.centos.org) (gcc version 4.8.5 20150623 (Red Hat 4.8.5-39) (GCC) )

**Comment by nlohmann** [comment]
(Sorry for having moved this into a discussion - it looks like a bug indeed).

Can you please execute

```cpp
std::cout<<j["a"].type_name()<<std::endl;
```

**Reply by Ymxia1993** [reply]
```
  std::ifstream file("Test.json");
  json j = json::parse(file);
  std::cout<<j["a"].type_name()<<std::endl;
  std::cout<<j["a"].is_number()<<std::endl;
```
I execute this  the results is:
```
number
0
```
type_name is number but when use is_number(), the result is 0, its really strange.

**Reply by nlohmann** [reply]
My bad - `type_name` will always return `number` for all number types.

Can you please output:

```cpp
std::cout<<j["a"].is_number()<<std::endl;
std::cout<<j["a"].is_number_integer()<<std::endl;
std::cout<<j["a"].is_number_unsigned()<<std::endl;
std::cout<<j["a"].is_number_float()<<std::endl;
```

**Reply by Ymxia1993** [reply]
is_number_unsigned() is correct 1, is_number_integer() is wrong 0 , is_number_float() is 0
